apidef/oas/authentication.go

```diff
@@ -647,18 +647,97 @@ type AuthenticationPlugin struct {
 	Path string `bson:"path" json:"path"` // required.
 	// RawBodyOnly if set to true, do not fill body in request or response object.
 	RawBodyOnly bool `bson:"rawBodyOnly,omitempty" json:"rawBodyOnly,omitempty"`
+	// IDExtractor  configures ID extractor with coprocess custom authentication.
+	IDExtractor *IDExtractor `bson:"idExtractor,omitempty" json:"idExtractor,omitempty"`
 }
 
 func (ap *AuthenticationPlugin) Fill(api apidef.APIDefinition) {
 	ap.FunctionName = api.CustomMiddleware.AuthCheck.Name
 	ap.Path = api.CustomMiddleware.AuthCheck.Path
 	ap.RawBodyOnly = api.CustomMiddleware.AuthCheck.RawBodyOnly
 	ap.Enabled = !api.CustomMiddleware.AuthCheck.Disabled
+	if ap.IDExtractor == nil {
+		ap.IDExtractor = &IDExtractor{}
+	}
+
+	ap.IDExtractor.Fill(api)
+	if ShouldOmit(ap.IDExtractor) {
+		ap.IDExtractor = nil
+	}
 }
 
 func (ap *AuthenticationPlugin) ExtractTo(api *apidef.APIDefinition) {
 	api.CustomMiddleware.AuthCheck.Disabled = !ap.Enabled
 	api.CustomMiddleware.AuthCheck.Name = ap.FunctionName
 	api.CustomMiddleware.AuthCheck.Path = ap.Path
 	api.CustomMiddleware.AuthCheck.RawBodyOnly = ap.RawBodyOnly
+
+	if ap.IDExtractor != nil {
+		ap.IDExtractor.ExtractTo(api)
+	}
+}
+
+type IDExtractorConfig struct {
+	// HeaderName is the header name to extract ID from.
+	HeaderName string `bson:"headerName,omitempty" json:"headerName,omitempty"`
+	// FormParamName is the form parameter name to extract ID from.
+	FormParamName string `bson:"formParamName,omitempty" json:"formParamName,omitempty"`
+	// Regexp is the regular expression to match ID.
+	Regexp string `bson:"regexp,omitempty" json:"regexp,omitempty"`
+	// RegexpMatchIndex is the index from which ID to be extracted after a match.
+	RegexpMatchIndex int `bson:"regexpMatchIndex,omitempty" json:"regexpMatchIndex,omitempty"`
+	// XPathExp is the xpath expression to match ID.
+	XPathExp string `bson:"xPathExp,omitempty" json:"xPathExp"`
+}
+
+func (id *IDExtractorConfig) Fill(api apidef.APIDefinition) {
+	id.HeaderName = api.CustomMiddleware.IdExtractor.ExtractorConfig.HeaderName
+	id.FormParamName = api.CustomMiddleware.IdExtractor.ExtractorConfig.FormParamName
+	id.Regexp = api.CustomMiddleware.IdExtractor.ExtractorConfig.RegexExpression
+	id.RegexpMatchIndex = api.CustomMiddleware.IdExtractor.ExtractorConfig.RegexMatchIndex
+	id.XPathExp = api.CustomMiddleware.IdExtractor.ExtractorConfig.XPathExpression
+}
+
+func (id *IDExtractorConfig) ExtractTo(api *apidef.APIDefinition) {
+	api.CustomMiddleware.IdExtractor.ExtractorConfig.HeaderName = id.HeaderName
```

Comment:
This looks to be nearly the full object definition... why not overwrite it with *val = T{}?